Support 'nicer' ways of writing affine handlers

dolio opened the pull request:
This PR adds support for a version of affine handlers that don't directly use the `Request` type. This is relatively common, since it's rather nice to write. The general form is something like:

``` unison
theHandler thunk =
  handle !thunk with cases
    { eff ... -> k } ->
      ...
      theHandler do
        ...
        k ...
```

That is, you have a recursive function taking "thunks" and handle the forced thunk within it.

The strategy for supporting these was to make the ANF optimizations more robust, and rewrite these handlers into the `Request` version that is easily recognizable. The optimizations used just eliminate indirections, too, so they end up speeding up these sorts of handlers. In the end, these handlers end up getting rewritten to perform (nearly) identically to the more stilted version with a separate 'handler loop' with a wrapper.

This also fixes an oversight with the last handler optimization that wasn't recognizing any affine handlers that required values to be built to call the continuation. So, if your recursive call looked like:

```unison
handle k () with myself
```

it wouldn't get recognized as an affine handler, because `k ()` was an opaque call like `callK123 k`. This has been fixed, at least for simple cases like this. I assume this was blocking many affine handlers from getting recognized.

The optimizations involved have been split into their own file, because they're getting kind of extensive. This also makes debugging them easier.

I've added some functions for debugging intermediate values that I found useful while developing this. I also fixed a bug with variable renaming in `ABT.Normalized`. I don't think it affected anything before, because nothing was doing any interesting renaming at this level.

Base tests and transcripts pass.

dolio committed:
Fix a variable renaming bug in normalized ABT

Renamed bound variables weren't being freshened against the right set
of other variables. It seems renaming was never used in a situation
where this mattered before, though.

dolio committed:
Split ANF optimizations into their own file. Improve affine handlers.

Inlining has been rewritten and extended. Some 'tail only' inlinings
are now recognized, which help to rewrite nicer affine handlers into
the originally recognized form.

Some 'peephole' optimizations have been added, which further rewrite
handlers into nicer forms, though they may apply in other scenarios
(improving performance slightly).

Now, two sorts of affine handlers that were previously unrecognized
are. One is handlers that require data values in the handler block:

    handle k () with h

These were previously blocked by havin an opaque function in the
runtime version of the 'handle' call.

Next, handlers based around 'thunks' are now supported, e.g.

    f th = handle !th with cases
      { eff <vs> -> k } ->
        ...
        f do k <us>

These are optimized into the more direct form, so should perform
almost identically.

dolio committed:
Add some additional tracing functions

dolio committed:
Break lines when displaying intermediate handle constructs

dolio committed:
Merge remote-tracking branch 'origin/trunk' into topic/nicer-affine-handlers

dolio committed:
automatically run ormolu

ceedubs commented:
Nice! I'll run the cloud tests with this when I get a chance.

aryairani commented:
@ceedubs go ahead and merge once you're sure it's passing — thanks

pchiusano commented:
I'm fine with delaying this until Dan implements the further optimizations. As is we don't expect this PR to speed up Cloud.

ceedubs commented:
I've confirmed that the Cloud integration tests work fine with this change. I have no preference on whether we merge this now or wait for further changes from Dan, so I'll leave that up to y'all.

dolio committed:
Rework optimization to include more affine handlers

This includes some extensive changes to improve overall optimization
and recognize more affine handler cases.

- Optimization information is now stored in the `CCache`. This is
  necessary because things like the inlining information are now based
  on the _optimized_ versions of definitions.

- The optimizer now accepts all the combinators at a time, and
  processes them in dependency order. This is necessary to base later
  optimizations on the optimized version of their dependencies.

- Inlinings are now classified into one of three categories:

    1. inline anywhere
    2. inline only in tail position
    3. don't inline

  The don't inline case is still useful, because other optimizations
  make use of the information.

- A peephole optimization has been added that will improve let
  bindings from 'indirect' to 'direct' if possible. This allows
  more subsequent inlining, because when primop wrappers are inlined,
  they are subsequently recognized as not being indirect calls.

- The update operation for affine handlers is now able to re-institute
  the handler into the current environment, allowing affine handlers
  where the resumption body might use the same handler. This is
  avoided when unnecessary, because it has some cost.

The new optimization framework conceivably allows for additional
optimization of mutually recursive functions, but not much along those
lines has been implemented yet (only special cases involved in 'nice'
self-recursive handlers).

dolio committed:
automatically run ormolu

dolio commented:
I've added some capabilities to the optimization here.

Now it should be possible to have handlers like:

``` unison
...
handle
   k (whatever x y z)
with myself ...
```

even if `whatever` is not simple enough to be optimized away (which would be a tall order). Essentially, you can have whatever you want in the `handle` body as long as the call to the continuation is _syntactically_ the last thing in the body (i.e. you _cannot_ just hand it off to a function that after some indirections eventually calls it in tail position). However, it will generally be _faster_ if you hoist your arbitrary function calls out of the handle body.

Some transcripts are changed, but I think it's actually the optimizations making them better (like `bug` gets inlined and now you see the function you called it from in the stack trace, instead of literally "bug").

I have a few corner cases to fix, but this can probably be tried out at this point.

dolio committed:
Fix rewriting combinators to descend under non-unary lets

dolio committed:
Condense optimization in U.R.Machine

dolio committed:
Add a note on indirect calls to `replaceLinearBody`

dolio committed:
Tweak some COMPLETE pragmas for ANF

They were good enough for most normal code, but there were some
corner cases where they were not quite complete.

dolio committed:
Adjust some affine optimizations

Don't worry about indirection for eliminating unused let bindings
in an affine handler.

Do local inlining of `y = x` bindings in affine handlers.

dolio committed:
Fix up tests

dolio committed:
Don't promote `x = y` bindings to direct

Seems harmless, but the mcode compiler will complain about it, and
that might be more complicated to fix right now

dolio committed:
Fix some issues causing cli integration test failures

Optimization info for builtin stuff wasn't getting stripped out. This
was causing the top level exception handler to be written out, and it
contains a `Discard` which we refuse to serialize.

---

Serializing optimization info involves inlined foreign functions,
which we previously refused to read back in. I've changed that, and
the scheme for serializing them.

They were never part of the public interchange format. The only reason
for supporting their serialization was for the scheme compiler. But
now it directly implements builtins, so it shouldn't need to know
about the wrappers, which is the only place they would occur.

I've switched to using numbers for foreign functions. It uses the
derived enum numbering, because this is just for standalone programs,
which don't have stable versioning (each new ucm release is a separate
compiled version). There is a flag passed to the serialization which
controls whether a foreign op occurrence causes an error. The error
should occur on all interchange level stuff, and only not occur for
the standalone serialization.

dolio committed:
Transcript changes

These are stack trace changes. Now calls to `bug` get inlined, so you
see the actual place that it's called, rather than just "bug".

dolio committed:
Fix tests

dolio committed:
Merge remote-tracking branch 'origin/trunk' into topic/nicer-affine-handlers

dolio committed:
automatically run ormolu

pchiusano reviewed (approved):
Dan, why don't you make a transcript that is self-contained and illustrates a few handlers in different forms that are all optimized by this PR. It'll be good both for testing and documentation purposes.

dolio committed:
Try adding a transcript that checks some affine handler performance

dolio committed:
Add some explanation to the affine handler transcript

dolio committed:
Merge remote-tracking branch 'refs/remotes/origin/topic/nicer-affine-handlers' into topic/nicer-affine-handlers

dolio commented:
I tried making a test case that would have enough of a speed up to consistently be detectable, even if it's not very realistic.

I did get one failure once locally, but I haven't been able to get it to fail a second time, so I'm not sure what happened. I think 2x speed up is fairly conservative for this test (having 20 irrelevant handlers to skip past).

dolio commented:
Okay, not good enough for CI, I guess.

dolio committed:
Make affine speedup test more reliable
